runtime/vm/jvminit.c

```diff
@@ -7896,3 +7939,162 @@ parseGlrOption(J9JavaVM* jvm, char* option)
+{
+	PORT_ACCESS_FROM_PORT((J9PortLibrary *)userData);
+
+	j9mem_free_memory(*(J9UTF8 **)entry);
+	return TRUE;
+}
+
+/**
+ * This should be called to clean up the ensureHashed configuration
+ *
+ * @param jvm pointer to J9JavaVM that can be used by the method
+ */
+static void
+cleanupEnsureHashedConfig(J9JavaVM* jvm)
+{
+	PORT_ACCESS_FROM_JAVAVM(jvm);
+
+	if (NULL != jvm->ensureHashedClasses) {
+		hashTableForEachDo(jvm->ensureHashedClasses, ensureHashedDoDelete, PORTLIB);
+		hashTableFree(jvm->ensureHashedClasses);
+	}
+	jvm->ensureHashedClasses = NULL;
+}
+
+/**
+ * This function parses a specific option
+ *
+ * @param option the option to be parsed
+ * @returns JNI_OK on success
+ */
+static UDATA
+parseEnsureHashedOption(J9JavaVM *jvm, char *option, BOOLEAN isAdd)
+{
+	PORT_ACCESS_FROM_JAVAVM(jvm);
+	char *className = option;
+	size_t classNameLength = strlen(className);
+	J9UTF8 *hashtableEntry;
+
+	if (0 == strcmp(option, "")) {
+		/* Empty class name */
+		j9nls_printf(PORTLIB, J9NLS_ERROR, J9NLS_VM_INVALID_EMPTY_OPTION, isAdd ? VMOPT_XXENABLEENSUREHASHED : VMOPT_XXDISABLEENSUREHASHED);
+		return J9VMDLLMAIN_FAILED;
+	}
+
+	if (NULL == jvm->ensureHashedClasses) {
+		jvm->ensureHashedClasses = hashTableNew(OMRPORT_FROM_J9PORT(PORTLIB),
+				J9_GET_CALLSITE(),
+				ENSUREHASHED_INITIAL_TABLE_SIZE,
+				sizeof(char*),
+				0,
+				0,
+				OMRMEM_CATEGORY_VM,
+				ensureHashedHashFn,
+				ensureHashedHashEqualFn,
+				NULL,
+				PORTLIB);
+
+		if (NULL == jvm->ensureHashedClasses) {
+			/* we failed the allocation */
+			j9nls_printf(PORTLIB, J9NLS_ERROR, J9NLS_ENSUREHASHED_CONFIG_OUT_OF_MEM);
+			return JNI_ENOMEM;
+		}
+	}
+
+	hashtableEntry = j9mem_allocate_memory(classNameLength + sizeof(J9UTF8_LENGTH(hashtableEntry)), OMRMEM_CATEGORY_VM);
+	if (NULL == hashtableEntry) {
+		j9nls_printf(PORTLIB, J9NLS_ERROR, J9NLS_ENSUREHASHED_CONFIG_OUT_OF_MEM);
+		return JNI_ENOMEM;
+	}
+	memcpy(J9UTF8_DATA(hashtableEntry), className, classNameLength);
+	J9UTF8_SET_LENGTH(hashtableEntry, (U_16)classNameLength);
+
+	if (TRUE == isAdd) {
+		/* -XX:+EnsureHashed: */
+		if (NULL == hashTableAdd(jvm->ensureHashedClasses, &hashtableEntry)) {
+			j9mem_free_memory(hashtableEntry);
+			j9nls_printf(PORTLIB, J9NLS_ERROR, J9NLS_ENSUREHASHED_CONFIG_OUT_OF_MEM);
+			return JNI_ENOMEM;
+		}
+	} else {
+		/* -XX:-EnsureHashed: */
+		hashTableRemove(jvm->ensureHashedClasses, &hashtableEntry);
+		j9mem_free_memory(hashtableEntry);
+	}
+
+	return JNI_OK;
+}
+
+/**
+ * This function parses a string containing ensureHashed options
+ *
+ * @param options string containing the options specified on the command line
+ * @returns JNI_OK on success
+ */
+static UDATA
+parseEnsureHashedConfig(J9JavaVM *jvm, char *options, BOOLEAN isAdd)
+{
+	UDATA result = JNI_OK;
+	char *nextOption = NULL;
+	char *cursor = options;
+	PORT_ACCESS_FROM_JAVAVM(jvm);
+
+	/* parse out each of the options */
+	while ((JNI_OK == result) && (NULL != (strstr(cursor, ",")))) {
```

Comment:
We have the option string and then allocate a new string in `scan_to_delim` and then allocate another copy (now with UTF8 header) in `parseEnsureHashedOption`.

This could be done without the extra allocation.  Given that lockwordconfig has the same behaviour, and this unlikely to be a heavily used option, I'm ok with leaving it as is.  Just think about where the allocations are in the code and if they can (and should) be avoided.